QAChat/Data_Processing/confluence_preprocessor.py

```diff
@@ -0,0 +1,248 @@
+        hash_object = hashlib.sha256()
+        hash_object.update(text.encode("utf-8"))
+        hash_value = hash_object.hexdigest()
+
+        return hash_value
+
+    def get_spaces(self):
+        start = 0
+        limit = 500
+        spaces = []
+
+        # Get all spaces starting from 0 to 500 (max of possible retrieved spaces via API)
+        # While there is no empty response, keep making requests
+        while True:
+            new_spaces = self.confluence.get_all_spaces(
+                start=start, limit=limit, expand=None
+            )
+            if new_spaces["size"] == 0:
+                break
+            spaces.extend(new_spaces["results"])
+            start += limit
+
+        # Exclude personal/user spaces, only include global spaces, and save necessary data
+        for space in spaces:
+            if space["type"] == "global":
+                if space["key"] not in self.restricted_spaces:
+                    self.spaces.append({"key": space["key"], "name": space["name"]})
+
+    def get_page_by_id(self, id):
+        # Retrieve page content by ID and extract the text from HTML
+        return self.get_text_from_html(
+            self.confluence.get_page_by_id(id, expand="body.storage")["body"][
+                "storage"
+            ]["value"]
+        )
+
+    def get_comments(self, id):
+        # Retrieve comments for a page and extract the text from HTML
+        comments = []
+        for i in self.confluence.get_page_child_by_type(
+            id, type="comment", start=None, limit=None, expand="body.storage"
+        ):
+            comments.append(self.get_text_from_html(i["body"]["storage"]["value"]))
+        return comments
+
+    def get_page_attacments(
+        self, id
+    ):  # TODO: Change this to use the one who implemented by the team
+        attachments_container = self.confluence.get_attachments_from_content(
+            page_id=id, start=0, limit=500
+        )
+        attachments = attachments_container["results"]
+        pdf_content = []
+        for attachment in attachments:
+            content = ""
+            fname = attachment["title"]
+            download_link = self.confluence.url + attachment["_links"]["download"]
+            r = requests.get(
+                download_link, auth=(self.confluence.username, self.confluence.password)
+            )
+            if r.status_code == 200:
+                #   with open(fname, "wb") as f:
+                #       for bits in r.iter_content():
+                #           f.write(bits)
+                pdf_file = io.BytesIO(r.content)
+                pdf_reader = PyPDF2.PdfReader(pdf_file)
+                num_pages = len(pdf_reader.pages)
+
+                for page_num in range(num_pages):
+                    page = pdf_reader.pages[page_num]
+                    text = page.extract_text()
+                    content += text
+                pdf_content.append(content)
```

Comment:
Why do you have this. The PDF Reading is implemented by Felix and me and only needs to be used. You just need to give the pdf to the function and your good to go.